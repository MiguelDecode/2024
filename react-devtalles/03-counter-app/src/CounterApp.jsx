import PropTypes from "prop-types";
import { useState } from "react";

function CounterApp({ value }) {
  const [counter, setCounter] = useState(value);

  console.log("Render");

  const handleAdd = () => setCounter(counter + 1);
  const handleSubstract = () => setCounter(counter - 1);
  const handleReset = () => setCounter(value);

  return (
    <>
      <h1>CounterApp</h1>
      <h2>{counter}</h2>

      <button aria-label="add" onClick={handleAdd}>+1</button>
      <button aria-label="substract" onClick={handleSubstract}>-1</button>
      <button aria-label="reset" id="btn-reset" onClick={handleReset}>Reset</button>
    </>
  );
}

CounterApp.propTypes = {
  value: PropTypes.number.isRequired,
};

export default CounterApp;
