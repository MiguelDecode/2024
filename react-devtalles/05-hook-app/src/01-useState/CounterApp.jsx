import { useState } from "react";

export const CounterApp = () => {
  const [state, setState] = useState({
    counter1: 10,
    counter2: 20,
    counter3: 30,
  });

  const { counter1, counter2, counter3 } = state;

  //   Care with objects of multiple levels and spread
  const handleAdd = () => {
    setState({ ...state, counter1: counter1 + 1 });
  };

  const handleSubstract = () => {
    setState({ ...state, counter1: counter1 - 1 });
  };

  return (
    <>
      <h1>Counter1: {counter1}</h1>
      <h1>Counter2: {counter2}</h1>
      <h1>Counter3: {counter3}</h1>

      <hr />

      <button className="btn btn-primary" onClick={handleAdd}>
        +1
      </button>
      <button className="btn btn-primary" onClick={handleSubstract}>
        -1
      </button>
    </>
  );
};
