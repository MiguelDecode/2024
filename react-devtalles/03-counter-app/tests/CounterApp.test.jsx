import { fireEvent, render, screen } from "@testing-library/react";
import CounterApp from "../src/CounterApp";

describe("Pruebas en el <CounterApp/>", () => {
  const initialValue = 100;

  test("Debe hacer match con el snapshot", () => {
    const { container } = render(<CounterApp value={initialValue} />);

    expect(container).toMatchSnapshot();
  });

  test("Debe de mostrar el valor inicial de 100", () => {
    render(<CounterApp value={initialValue} />);

    expect(screen.getByText(100)).toBeTruthy();
    expect(screen.getByRole("heading", { level: 2 }).textContent).toContain(
      "100"
    );
  });

  test("Debe de incrementar con el botón +1", () => {
    render(<CounterApp value={initialValue} />);

    fireEvent.click(screen.getByText("+1"));

    expect(screen.getByText("+1")).toBeTruthy();
    expect(screen.getByRole("heading", { level: 2 }).textContent).toContain(
      "101"
    );
  });

  test("Debe de decrementar con el botón de -1", () => {
    render(<CounterApp value={initialValue} />);

    fireEvent.click(screen.getByText("-1"));

    expect(screen.getByText("-1")).toBeTruthy();
    expect(screen.getByRole("heading", { level: 2 }).textContent).toContain(
      "99"
    );
  });

  test("Debe de funcionar el botón de reset", () => {
    render(<CounterApp value={initialValue} />);

    fireEvent.click(screen.getByText("-1"));
    fireEvent.click(screen.getByText("-1"));
    fireEvent.click(screen.getByText("-1"));

    fireEvent.click(screen.getByText("Reset"));

    expect(screen.getByText("Reset")).toBeTruthy();
    expect(screen.getByText("100")).toBeTruthy();
    expect(screen.getByRole("button", { name: "reset" })).toBeTruthy();
    expect(screen.getByRole("heading", { level: 2 }).textContent).toContain(
      "100"
    );
  });
});
