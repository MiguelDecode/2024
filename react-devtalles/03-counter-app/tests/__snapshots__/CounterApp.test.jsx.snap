// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <CounterApp/> Debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    CounterApp
  </h1>
  <h2>
    100
  </h2>
  <button
    aria-label="add"
  >
    +1
  </button>
  <button
    aria-label="substract"
  >
    -1
  </button>
  <button
    aria-label="reset"
    id="btn-reset"
  >
    Reset
  </button>
</div>
`;
